File: gherkin/java/src/test/java/io/cucumber/gherkin/GherkinDocumentBuilderTest.java
package io.cucumber.gherkin;

import io.cucumber.messages.IdGenerator;
import io.cucumber.messages.types.Comment;
import io.cucumber.messages.types.FeatureChild;
import io.cucumber.messages.types.GherkinDocument;
import io.cucumber.messages.types.Pickle;
import io.cucumber.messages.types.TableRow;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class GherkinDocumentBuilderTest {
    private final IdGenerator idGenerator = new IncrementingIdGenerator();

    @Test
    public void is_reusable() {
        Parser<GherkinDocument> parser = new Parser<>(new GherkinDocumentBuilder(idGenerator, "test.feature"));
        TokenMatcher matcher = new TokenMatcher();

        GherkinDocument d1 = parser.parse("Feature: 1", matcher, "1.feature");
        GherkinDocument d2 = parser.parse("Feature: 2", matcher, "2.feature");

        assertEquals("1", d1.getFeature().get().getName());
        assertEquals("2", d2.getFeature().get().getName());
    }

    @Test
    public void parses_rules() {
        Parser<GherkinDocument> parser = new Parser<>(new GherkinDocumentBuilder(idGenerator, "test.feature"));
        String data = "" +
                "Feature: Some rules\n" +
                "\n" +
                "  Background:\n" +
                "    Given fb\n" +
                "\n" +
                "  Rule: A\n" +
                "    The rule A description\n" +
                "\n" +
                "    Background:\n" +
                "      Given ab\n" +
                "\n" +
                "    Example: Example A\n" +
                "      Given a\n" +
                "\n" +
                "  Rule: B\n" +
                "    The rule B description\n" +
                "\n" +
                "    Example: Example B\n" +
                "      Given b";
        GherkinDocument doc = parser.parse(data, "test.feature");

        List<FeatureChild> children = doc.getFeature().get().getChildren();
        assertEquals(3, children.size());

        IdGenerator idGenerator = new IncrementingIdGenerator();
        PickleCompiler pickleCompiler = new PickleCompiler(idGenerator);
        List<Pickle> pickles = pickleCompiler.compile(doc, "hello.feature");
        assertEquals(2, pickles.size());

        assertEquals(3, pickles.get(0).getSteps().size());

        assertEquals(2, pickles.get(1).getSteps().size());
    }

    @Test
    public void parses_just_comments() {
        Parser<GherkinDocument> parser = new Parser<>(new GherkinDocumentBuilder(idGenerator, "test.feature"));
        GherkinDocument doc = parser.parse("# Just a comment", "test.feature");
        List<Comment> children = doc.getComments();
        assertEquals(1, children.size());
    }

    @Test
    public void sets_empty_table_cells() {
        Parser<GherkinDocument> parser = new Parser<>(new GherkinDocumentBuilder(idGenerator, "test.feature"));
        GherkinDocument doc = parser.parse("" +
                "Feature:\n" +
                "  Scenario:\n" +
                "    Given a table\n" +
                "      |a||b|",
                "test.feature"
        );
        TableRow row = doc.getFeature().get().getChildren().get(0).getScenario().get().getSteps().get(0).getDataTable().get().getRows().get(0);
        assertEquals("a", row.getCells().get(0).getValue());
        assertEquals("", row.getCells().get(1).getValue());
        assertEquals("b", row.getCells().get(2).getValue());
    }
}

File: gherkin/dart/test/GherkinDocumentBuilder_test.dart
import 'package:gherkin/ast.dart';
import 'package:gherkin/language.dart';
import 'package:gherkin/parser.dart';
import 'package:gherkin/pickles.dart';
import 'package:test/test.dart';

void main()
{
    final languages = loadGherkinLanguagesFromJsonAsset();
    final dialectProvider = GherkinDialectProvider(languages);
    final idGenerator = IdGenerator.incrementingGenerator;
    final builder = GherkinDocumentBuilder(idGenerator);
    final parser = Parser<GherkinDocument>(builder);
    final matcher = TokenMatcher(dialectProvider);

    test('Is reusable', () {
        var tokenScanner = StringTokenScanner('Feature: 1');
        var d1 = parser.parse(tokenScanner, matcher);

        tokenScanner = StringTokenScanner('Feature: 2');
        var d2 = parser.parse(tokenScanner, matcher);

        expect(d1.feature.name, '1');
        expect(d2.feature.name, '2');
    } );

    test('Parses rules', () {
        var data =
            'Feature: Some rules\n'
            '\n'
            '  Background:\n'
            '    Given fb\n'
            '\n'
            '  Rule: A\n'
            '    The rule A description\n'
            '\n'
            '    Background:\n'
            '      Given ab\n'
            '\n'
            '    Example: Example A\n'
            '      Given a\n'
            '\n'
            '  Rule: B\n'
            '    The rule B description\n'
            '\n'
            '    Example: Example B\n'
            '      Given b';

        var tokenScanner = StringTokenScanner(data);
        var doc = parser.parse(tokenScanner, matcher);

        var children = doc.feature.children.toList();
        expect(children.length, 3);

        var pickleCompiler = PickleCompiler(idGenerator);
        var pickles = pickleCompiler.compile(doc, 'hello.feature');
        expect(2, pickles.length);

        expect(3, pickles.first.steps.length);

        expect(2, pickles.elementAt(1).steps.length);
    } );

    test('Parses just comments', () {
        final data = '# Just a comment';
        var tokenScanner = StringTokenScanner(data);
        var doc = parser.parse(tokenScanner, matcher);
        var children = doc.comments;
        expect(children.length, 1);
    } );

    test('Sets empty table cells', () {
        final data =
            'Feature:\n'
            '  Scenario:\n'
            '    Given a table\n'
            '      |a||b|';
        var tokenScanner = StringTokenScanner(data);
        var doc = parser.parse(tokenScanner, matcher);

        var row = doc.feature
            .children.first
            .scenario
            .steps.first
            .dataTable
            .rows.first;
        expect('a', row.cells.first.value);
        expect('', row.cells.elementAt(1).value);
        expect('b', row.cells.elementAt(2).value);
    } );
}
